src/cirrus/release.py

```diff
@@ -503,6 +538,117 @@ def upload_release(opts):
+            'release', 'wait_on_ci_interval', 2)
+        )
+
+    repo_dir = os.getcwd()
+    curr_branch = get_active_branch(repo_dir)
+    expected_branch = release_branch_name(config)
+
+    if curr_branch.name != expected_branch:
+        msg = (
+            "Not on the expected release branch according "
+            "to cirrus.conf\n Expected:{0} but on {1}"
+        ).format(expected_branch, curr_branch)
+        LOGGER.error(msg)
+        raise RuntimeError(msg)
+    # merge in release branches and tag, push to remote
+    tag = config.package_version()
+    master = config.gitflow_master_name()
+    develop = config.gitflow_branch_name()
+
+    # merge release branch into master
+    LOGGER.info("Tagging and pushing {0}".format(tag))
+    checkout_and_pull(repo_dir, master)
+    master_merge_sha = merge(repo_dir, master, expected_branch)
+
+    if not opts.test:
+        LOGGER.info("pushing to remote...")
+        if release_config['wait_on_ci']:
+            # if wait_on_ci is set and we have gotten to this point,
+            # tests pass on the release branch, so we can tell the
+            # remote the good news.
+            current_branch_mark_status(repo_dir, 'success')
+        push(repo_dir)
+    do_push = not opts.test
+    tag_release(repo_dir, tag, master, push=do_push)
+    if opts.merge_comment is not None:
+        #
+        # add the provided comment to the last commit on master
+        #
+        path = 'cirrus.conf'  # TODO, make this configurable
+        comment_on_sha(
+            config.organization_name(),
+            config.package_name(),
+            opts.merge_comment,
+            master_merge_sha,
+            path)
+
+    # Merge release branch back to develop
+    # push to develop
+    LOGGER.info("Merging back to develop...")
+    checkout_and_pull(repo_dir, develop)
+    if release_config['wait_on_ci']:
+        # @HACK If we are doing CI, our branch will be "out of date",
+        # so we update it by merging from master (not gitflow, I
+        # know), and then waiting for CI status on the remote.
+        develop_merge_sha = merge(repo_dir, develop, master)
+        wait_on_gh_status(
+            develop,
+            timeout=release_config['wait_on_ci_timeout'],
+            interval=release_config['wait_on_ci_interval']
+        )
+    else:
+        develop_merge_sha = merge(repo_dir, develop, expected_branch)
+
+    if not opts.test:
+        LOGGER.info("pushing to remote...")
+        if release_config['wait_on_ci']:
+            # if wait_on_ci is set and we have gotten to this point,
+            # tests pass on the release branch, so we can tell the
+            # remote the good news.
+            current_branch_mark_status(repo_dir, 'success')
+        push(repo_dir)
+
+    if opts.merge_comment is not None:
+        #
+        # add the provided comment to the last commit on develop
+        #
+        path = 'cirrus.conf'  # TODO, make this configurable
+        comment_on_sha(
```

Comment:
Commenting after we attempt to do the push won't unblock us though, right? We'd need to drop in the comment on a remote version that is awaiting merging, so that the push above doesn't error out.